docs/source/tutorial/arrays.rst

```diff
@@ -0,0 +1,133 @@
+======
+Arrays
+======
+
+To get the full power out of a C++ extension, you will often need to pass arrays of data between Python and C++.
+Libpy has native support for integrating with numpy, the most popular ndarray library for Python.
+
+Libpy supports receiving arrays as views so that no data needs to be copied.
+Libpy array views can also be const to guarantee that the underlying data isn't mutated.
+Libpy also supports creating Numpy arrays as views over C++ containers without copying the underlying data.
+
+``py::array_view``
+==================
+
+Libpy can accept numpy arrays, or generally any buffer-like object, through a :cpp:struct:`py::ndarray_view`.
+:cpp:struct:`py::ndarray_view` is a template type which takes as a parameter the C++ type of the elements of the array and the number of dimensions.
+For example: ``py::ndarray_view<std::int32_t, 3>`` is a view of a 3d array of signed 32 bit integers.
+The type of the elements of a :cpp:struct:`py::ndarray_view` are fixed at compile time, but the shape is determined at runtime.
+
+As a convenience, :cpp:struct:`py::array_view` is an alias of :cpp:struct:`py::ndarray_view` for 1 dimensional arrays.
+
+Shape and Strides
+-----------------
+
+Like numpy, an array view is composed of three parts:
+
+- shape :: ``std::array<std::size_t>``
+- strides :: ``std::array<std::int64_t>``
+- buffer :: ``(const) std::byte*``
+
+The shape array contains the number of elements along each axis.
+For example: ``{2, 3}`` would be an array with 2 rows and 3 columns.
+
+The strides array contains the number of bytes needed to move one step along each axis.
+For example: given a ``{2, 3}`` shaped array of 4 byte elements, then strides of ``{12, 4}`` would be a C-contiguous array because the rows are contiguous.
+Given the same ``{2, 3}`` shaped array of 4 byte elements, then strides of ``{4, 8}`` would be a Fortran-contiguous array because the rows are contiguous.
+
+Non-contiguous views
+--------------------
+
+Array views do not need to view contiguous arrays.
+For example, given a C-contiguous ``{4, 5}`` array of 2 byte values, we could take a view of first column by producing an array view with strides ``{10}``.
+
+Simple Array Input
+==================
+
+Let's write function to sum an array:
+
+.. code-block:: c++
```

Comment:
This could just include from `libpy_tutorials/arrays.cc`

Reply:
I am not sure I like the includes. If I change the code, I need to go change the lines in source, but probably won't remember to do that so the docs are just broken.

Reply:
But this way you're only changing the underlying code in one place, rather than needing to make x updates anytime we change tutorial content. 

Plus if  everything is referring to source we compile it is more likely to be correct.